src/ucx_rma_plugin.c

```diff
@@ -0,0 +1,1062 @@
+
+  rkey_buf = malloc(rkey_buf_size);
+  if (rkey_buf == NULL) {
+    return ncclSystemError;
+  }
+  NCCLCHECK(socketReceive(comm->super.fd, rkey_buf, rkey_buf_size));
+  NCCLCHECK(socketReceive(comm->super.fd, &comm->rem_fifo.addr, sizeof(uint64_t)));
+  NCCLCHECK(socketReceive(comm->super.fd, &comm->rem_am_id, sizeof(int)));
+  UCXCHECK(ucp_ep_rkey_unpack(comm->ep, rkey_buf, &comm->rem_fifo.rkey));
+  free(rkey_buf);
+
+  if (nccl_p2p_gdr_support(l_comm->dev) == ncclSuccess) {
+    comm->super.gpuFlush.enabled = 1;
+  }
+
+  if (comm->super.gpuFlush.enabled) {
+    ucp_worker_attr_t attr;
+    ucp_ep_params_t   ep_params;
+
+    attr.field_mask    = UCP_WORKER_ATTR_FIELD_ADDRESS |
+                         UCP_WORKER_ATTR_FIELD_ADDRESS_FLAGS;
+    attr.address_flags = UCP_WORKER_ADDRESS_FLAG_NET_ONLY;
+
+    UCXCHECK(ucp_worker_query(comm->super.worker, &attr));
+    ep_params.field_mask = UCP_EP_PARAM_FIELD_REMOTE_ADDRESS;
+    ep_params.address    = attr.address;
+    UCXCHECK(ucp_ep_create(comm->super.worker, &ep_params,
+                           &comm->super.gpuFlush.flush_ep));
+
+    free(attr.address);
+  }
+  comm->super.num_mh = 0;
+  *recv_comm = comm;
+
+  return ncclSuccess;
+}
+
+#define REG_ALIGN (4096)
+ncclResult_t nccl_ucx_rma_regmr(void* comm, void* data, int size, int type,
+                                void** mhandle)
+{
+  nccl_ucx_rma_ctx_t   *ctx = (nccl_ucx_rma_ctx_t*)comm;
+  uint64_t             addr = (uint64_t)data;
+  ucp_mem_map_params_t mmap_params;
+  ucx_rma_mhandle_t    *mh;
+  uint64_t             reg_addr, reg_size;
+  
+  reg_addr = addr & (~(REG_ALIGN - 1));
+  reg_size = addr + size - reg_addr;
+  reg_size = ROUNDUP(reg_size, REG_ALIGN);
+
+  mmap_params.field_mask = UCP_MEM_MAP_PARAM_FIELD_ADDRESS |
+                           UCP_MEM_MAP_PARAM_FIELD_LENGTH; 
+  mmap_params.address    = (void*)reg_addr;
+  mmap_params.length     = reg_size;
+  
+  mh = (ucx_rma_mhandle_t*)malloc(sizeof(ucx_rma_mhandle_t));
+  if (mh == NULL) {
+    return ncclSystemError;
+  }
+
+  UCXCHECK(ucp_mem_map(ctx->ctx, &mmap_params, &mh->ucp_memh));
+  UCXCHECK(ucp_rkey_pack(ctx->ctx, mh->ucp_memh, &mh->rkey_buf,
+                         &mh->rkey_buf_size));
+
+  if (ctx->gpuFlush.enabled) {
+    UCXCHECK(ucp_ep_rkey_unpack(ctx->gpuFlush.flush_ep, mh->rkey_buf, &mh->rkey));
+  }
+  
+  mh->index = ctx->num_mh;
+  ctx->mh[ctx->num_mh] = mh;
+  ctx->num_mh++;
+  *mhandle = mh;
+
+  return ncclSuccess;
+}
+
+ncclResult_t nccl_ucx_rma_deregmr(void* comm, void* mhandle)
+{
+  nccl_ucx_rma_ctx_t *ctx = (nccl_ucx_rma_ctx_t*)comm;
+  ucx_rma_mhandle_t  *mh  = (ucx_rma_mhandle_t*)mhandle;
+
+  if (ctx->gpuFlush.enabled) {
+      ucp_rkey_destroy(mh->rkey);
+  }
+  ucp_mem_unmap(ctx->ctx, mh->ucp_memh);
+  free(mh);
+
+  return ncclSuccess;
+}
+
+ncclResult_t ucx_rma_get_request(nccl_ucx_rma_request_t* reqs, int* req_id)
+{
+  nccl_ucx_rma_request_t *r;
+  int i;
+
+  for (i = 0; i < MAX_REQUESTS; i++) {
+    r = reqs + i;
```

Comment:
array search seems  very inefficient. may be we need to maintain a free list and used list